### PR TITLE
Ensure min/max value generation in the segment metadata. 

### PR DESCRIPTION
**Pull request by abhioncbr**
Labels:
`enhancement`
 
- ensuring that the min/max value is in the segment metadata. 
- trimmed the string/bytes value if exceeding the length limit. Added the test for the change.

As per the [issue](https://github.com/apache/pinot/issues/10793)

cc: @Jackie-Jiang

### CODE REVIEW

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentColumnarIndexCreator.java [hunk 1]
```diff
@@ -570,12 +570,14 @@ public static void addColumnMetadataInfo(PropertiesConfiguration properties, Str
...
-      properties.setProperty(getKeyFor(column, MIN_VALUE), minValue);
+      properties.setProperty(getKeyFor(column, MIN_VALUE),
+          minValue.substring(0, Math.min(minValue.length(), METADATA_PROPERTY_LENGTH_LIMIT)));
     } else {
       properties.setProperty(getKeyFor(column, MIN_MAX_VALUE_INVALID), true);
     }
     if (isValidPropertyValue(maxValue)) {
-      properties.setProperty(getKeyFor(column, MAX_VALUE), maxValue);
+      properties.setProperty(getKeyFor(column, MAX_VALUE),
+          maxValue.substring(0, Math.min(maxValue.length(), METADATA_PROPERTY_LENGTH_LIMIT)));
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentColumnarIndexCreator.java:580]: We cannot directly use substring as the max value because it is smaller than the actual max value and can cause false positive pruning. We want to pick a string that is larger than the actual max value

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/columnminmaxvalue/ColumnMinMaxValueGenerator.java [hunk 2]
```diff
@@ -119,65 +121,126 @@ private List<String> getColumnsToAddMinMaxValue() {
...
+            SegmentColumnarIndexCreator.addColumnMinMaxValueInfo(_segmentProperties, columnName,
+                bytesDictionary.getStringValue(0), bytesDictionary.getStringValue(length - 1));
+          }
+          break;
+        default:
+          throw new IllegalStateException("Unsupported data type: " + dataType + " for column: " + columnName);
+      }
+    } else {
+      // setting min/max for non-dictionary columns.
+      PinotDataBuffer forwardBuffer = _segmentWriter.getIndexFor(columnName, StandardIndexes.forward());
```
Comment (abhioncbr) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/columnminmaxvalue/ColumnMinMaxValueGenerator.java:184]: @Jackie-Jiang Can you please confirm my approach?

I am creating the `PinotDataBuffer` based on the forward index and then reading the by iterating over the doc.

Comment (Jackie-Jiang) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/columnminmaxvalue/ColumnMinMaxValueGenerator.java:184]: The approach is correct. We want to loop over all values from the forward index and gather the min/max value

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/columnminmaxvalue/ColumnMinMaxValueGenerator.java [hunk 3]
```diff
@@ -119,65 +121,126 @@ private List<String> getColumnsToAddMinMaxValue() {
...
+          try (BytesDictionary bytesDictionary = new BytesDictionary(dictionaryBuffer, length,
+              columnMetadata.getColumnMaxLength())) {
+            SegmentColumnarIndexCreator.addColumnMinMaxValueInfo(_segmentProperties, columnName,
+                bytesDictionary.getStringValue(0), bytesDictionary.getStringValue(length - 1));
+          }
+          break;
+        default:
+          throw new IllegalStateException("Unsupported data type: " + dataType + " for column: " + columnName);
+      }
+    } else if (_segmentMetadata.getSchema().getFieldSpecFor(columnName).getFieldType() == FieldSpec.FieldType.METRIC) {
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/columnminmaxvalue/ColumnMinMaxValueGenerator.java:182]: Why metric only?

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/columnminmaxvalue/ColumnMinMaxValueGenerator.java [hunk 4]
```diff
@@ -119,65 +121,126 @@ private List<String> getColumnsToAddMinMaxValue() {
...
+              DataType.DOUBLE); ChunkReaderContext readerContext = rawIndexReader.createContext()) {
+            Double[] minMaxValue = {Double.MAX_VALUE, Double.MIN_VALUE};
+            for (int docs = 0; docs < columnMetadata.getTotalDocs(); docs++) {
+              minMaxValue = getMinMaxValue(minMaxValue, rawIndexReader.getDouble(docs, readerContext));
+            }
+            SegmentColumnarIndexCreator.addColumnMinMaxValueInfo(_segmentProperties, columnName,
+                String.valueOf(minMaxValue[0]), String.valueOf(minMaxValue[1]));
+          }
+          break;
+        default:
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/columnminmaxvalue/ColumnMinMaxValueGenerator.java:230]: I think we also want to support `STRING` and `BYTES`

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentColumnarIndexCreator.java (hunk 1 above)
Comment (Jackie-Jiang) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentColumnarIndexCreator.java:580]: Seems we still have invalid values. In that case, that's not fix the length issue now

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/columnminmaxvalue/ColumnMinMaxValueGenerator.java (hunk 3 above)
Comment (abhioncbr) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/columnminmaxvalue/ColumnMinMaxValueGenerator.java]: I have made the change. However, do we also need to support MV fields? If yes, the max and min will be the max and min values seen in any array or something else? Thanks

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentColumnarIndexCreator.java (hunk 1 above)
Comment (Jackie-Jiang) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentColumnarIndexCreator.java]: I'd suggest creating a separate PR to address the invalid value issue if we do want to solve it. This PR can focus on generating min/max value for raw index column

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/columnminmaxvalue/ColumnMinMaxValueGenerator.java (hunk 3 above)
Comment (Jackie-Jiang) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/columnminmaxvalue/ColumnMinMaxValueGenerator.java]: Good point. For MV fields, we want to collect the min/max value of all the elements within the array. I.e. loop over the values and collect the min/max

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentColumnarIndexCreator.java (hunk 1 above)
Comment (abhioncbr) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentColumnarIndexCreator.java]: Sure. I am going to raise a separate PR for invalid values. Thanks

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentColumnarIndexCreator.java [hunk 5]
```diff
@@ -83,7 +83,7 @@ public class SegmentColumnarIndexCreator implements SegmentCreator {
   // TODO Refactor class name to match interface name
   private static final Logger LOGGER = LoggerFactory.getLogger(SegmentColumnarIndexCreator.class);
   // Allow at most 512 characters for the metadata property
-  private static final int METADATA_PROPERTY_LENGTH_LIMIT = 512;
+  static final int METADATA_PROPERTY_LENGTH_LIMIT = 512;
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentColumnarIndexCreator.java:86]: Let's revert this change and we can discuss it under #10990

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/columnminmaxvalue/ColumnMinMaxValueGenerator.java [hunk 6]
```diff
@@ -119,65 +123,153 @@ private List<String> getColumnsToAddMinMaxValue() {
 
   private boolean needAddColumnMinMaxValueForColumn(String columnName) {
     ColumnMetadata columnMetadata = _segmentMetadata.getColumnMetadataFor(columnName);
-    return columnMetadata.hasDictionary() && columnMetadata.getMinValue() == null
+    return columnMetadata.getMinValue() == null
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/columnminmaxvalue/ColumnMinMaxValueGenerator.java:126]: (nit) reformat

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/columnminmaxvalue/ColumnMinMaxValueGenerator.java [hunk 7]
```diff
@@ -119,65 +123,153 @@ private List<String> getColumnsToAddMinMaxValue() {
...
+          }
+          break;
+        case DOUBLE:
+          try (DoubleDictionary doubleDictionary = new DoubleDictionary(dictionaryBuffer, length)) {
+            SegmentColumnarIndexCreator.addColumnMinMaxValueInfo(_segmentProperties, columnName,
+                doubleDictionary.getStringValue(0), doubleDictionary.getStringValue(length - 1));
+          }
+          break;
+        case STRING:
+          try (StringDictionary stringDictionary = new StringDictionary(dictionaryBuffer, length,
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/columnminmaxvalue/ColumnMinMaxValueGenerator.java:168]: (code format) Does this follow [Pinot Style](https://docs.pinot.apache.org/developers/developers-and-contributors/code-setup#setup-ide)?

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/columnminmaxvalue/ColumnMinMaxValueGenerator.java [hunk 8]
```diff
@@ -119,65 +123,153 @@ private List<String> getColumnsToAddMinMaxValue() {
...
+      // setting min/max for non-dictionary columns.
+      PinotDataBuffer forwardBuffer = _segmentWriter.getIndexFor(columnName, StandardIndexes.forward());
+      switch (dataType) {
+        case INT:
+          try (FixedByteChunkSVForwardIndexReader rawIndexReader = new FixedByteChunkSVForwardIndexReader(forwardBuffer,
+              DataType.INT); ChunkReaderContext readerContext = rawIndexReader.createContext()) {
+            Integer[] minMaxValue = {Integer.MAX_VALUE, Integer.MIN_VALUE};
+            for (int docs = 0; docs < columnMetadata.getTotalDocs(); docs++) {
+              minMaxValue = getMinMaxValue(minMaxValue, rawIndexReader.getInt(docs, readerContext));
+            }
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/columnminmaxvalue/ColumnMinMaxValueGenerator.java:194]: Suggest directly comparing the primitive value to avoid the boxing/unboxing. Boxing/unboxing has quite expensive overhead. Same for other types
```suggestion
            int min = Integer.MAX_VALUE;
            int max = Integer.MIN_VALUE;
            int numDocs = columnMetadata.getTotalDocs();
            for (int i = 0; i < numDocs; i++) {
              int value = rawIndexReader.getInt(docs, readerContext);
              min = Math.min(min, value);
              max = Math.max(max, value);
            }
```

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/columnminmaxvalue/ColumnMinMaxValueGenerator.java (hunk 7 above)
Comment (abhioncbr) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/columnminmaxvalue/ColumnMinMaxValueGenerator.java:171]: I do not see any linting issues for this; it was like this before. But If needed, I can make the changes. Thanks

**Comment by abhioncbr** [comment]
@Jackie-Jiang, please have a look. Thanks

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/columnminmaxvalue/ColumnMinMaxValueGenerator.java (hunk 7 above)
Comment (Jackie-Jiang) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/columnminmaxvalue/ColumnMinMaxValueGenerator.java:171]: Nvm, I checked and it is the correct format. I somehow misread it

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/columnminmaxvalue/ColumnMinMaxValueGenerator.java [hunk 9]
```diff
@@ -119,65 +126,257 @@ private List<String> getColumnsToAddMinMaxValue() {
...
+      boolean isSingleValueField = _segmentMetadata.getSchema().getFieldSpecFor(columnName).isSingleValueField();
+      PinotDataBuffer forwardBuffer = _segmentWriter.getIndexFor(columnName, StandardIndexes.forward());
+      switch (dataType) {
+        case INT: {
+          int min = Integer.MAX_VALUE;
+          int max = Integer.MIN_VALUE;
+          if (isSingleValueField) {
+            FixedByteChunkSVForwardIndexReader rawIndexReader = new FixedByteChunkSVForwardIndexReader(forwardBuffer,
+                DataType.INT);
+            ChunkReaderContext readerContext = rawIndexReader.createContext();
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/columnminmaxvalue/ColumnMinMaxValueGenerator.java:199]: We want to close both the reader and the context. You may use the try-with-resource block
```suggestion
            try (FixedByteChunkSVForwardIndexReader rawIndexReader = new FixedByteChunkSVForwardIndexReader(forwardBuffer,
                DataType.INT);
              ChunkReaderContext readerContext = rawIndexReader.createContext()
            ) {
```

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/columnminmaxvalue/ColumnMinMaxValueGenerator.java [hunk 10]
```diff
@@ -119,65 +126,257 @@ private List<String> getColumnsToAddMinMaxValue() {
...
+            for (int docs = 0; docs < numDocs; docs++) {
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/columnminmaxvalue/ColumnMinMaxValueGenerator.java:200]: (nit) use `i` or `docId`

**Review by Jackie-Jiang — APPROVED** [review]
LGTM otherwise

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/columnminmaxvalue/ColumnMinMaxValueGenerator.java (hunk 4 above)
Comment (Jackie-Jiang) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/columnminmaxvalue/ColumnMinMaxValueGenerator.java:376]: @jadami10 Good catch. The intention is to handle all data types, or we will run this generator every time loading the segment. Clearly `BIG_DECIMAL` was left unhandled. It was actually not introduced in this PR as this PR only extends the handling to raw index.
+1 on making it more robust, and adding more tests to cover all these scenarios.

Comment (jadami10) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/columnminmaxvalue/ColumnMinMaxValueGenerator.java:376]: I think we're somewhat uniquely affected because of our use of metric aggregation which requires noDictionary.

previously, this code had
```
if (!columnMetadata.hasDictionary() || columnMetadata.getMinValue() != null
        || columnMetadata.getMaxValue() != null) {
      return;
    }
```

this PR removed the `hasDictionary` check and extended the min/max populating to those columns. But it also carried over the exception. Even for the previous branch, do we need to throw an exception for unsupported data types? It seems we can just not set the min/max.
